File: backend/app/schemas/tracking.py
"""Vehicle tracking data contracts - driver location reporting and admin fleet
monitoring. Additive to app/schemas/fleet.py (which gains a `tracking_code` per
vehicle route and a `planning_session_id` at the top level - see FleetRouteResponse).
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel

from app.schemas.routing import Coordinate


class LocationPingRequest(BaseModel):
    coordinate: Coordinate


class AssignedRoute(BaseModel):
    """What a driver's app shows after entering their tracking code: their
    assigned stops and route geometry, exactly as planned."""

    tracking_code: str
    vehicle_type: str
    stop_names: list[str]
    geometry: list[Coordinate]
    distance_meters: float
    duration_seconds: float


class VehicleTrackingStatus(BaseModel):
    """One vehicle's live status - used for both the driver's own map and each
    entry in the admin's fleet-wide view."""

    tracking_code: str
    vehicle_index: int
    vehicle_type: str
    stop_names: list[str]
    geometry: list[Coordinate]
    planned_distance_meters: float
    current_location: Coordinate | None = None
    last_ping_at: datetime | None = None
    distance_travelled_meters: float = 0.0
    is_off_route: bool = False
    off_route_distance_meters: float | None = None


class FleetTrackingOverview(BaseModel):
    job_id: str
    vehicles: list[VehicleTrackingStatus]

File: android/app/src/main/kotlin/com/qtrace/app/domain/model/Tracking.kt
package com.qtrace.app.domain.model

/** Mirrors backend/app/schemas/tracking.py::AssignedRoute - what a driver's app shows after
 * entering their tracking code. */
data class AssignedRoute(
    val trackingCode: String,
    val vehicleType: String,
    val stopNames: List<String>,
    val geometry: List<Coordinate>,
    val distanceMeters: Double,
    val durationSeconds: Double,
)

/** Mirrors backend/app/schemas/tracking.py::VehicleTrackingStatus - one vehicle's live status,
 * used by both the driver's own map and each entry in the admin's fleet-wide view. */
data class VehicleTrackingStatus(
    val trackingCode: String,
    val vehicleIndex: Int,
    val vehicleType: String,
    val stopNames: List<String>,
    val geometry: List<Coordinate>,
    val plannedDistanceMeters: Double,
    val currentLocation: Coordinate? = null,
    val lastPingAt: String? = null,
    val distanceTravelledMeters: Double = 0.0,
    val isOffRoute: Boolean = false,
    val offRouteDistanceMeters: Double? = null,
)

/** Mirrors backend/app/schemas/tracking.py::FleetTrackingOverview - the admin fleet-wide view. */
data class FleetTrackingOverview(
    val jobId: String,
    val vehicles: List<VehicleTrackingStatus>,
)
